cores/arduino/stm32/LowPower.c

```diff
@@ -0,0 +1,289 @@
+  if (__HAL_PWR_GET_FLAG(PWR_FLAG_SB) != RESET)
+  {
+    /* Clear Standby flag */
+    __HAL_PWR_CLEAR_FLAG(PWR_FLAG_SB);
+  }
+
+  /* Clear all related wakeup flags */
+  __HAL_PWR_CLEAR_FLAG(PWR_FLAG_WU);
+}
+
+/**
+  * @brief  Configure a pin as wakeup source if compatible.
+  * @param  pin: pin to configure
+  * @param  mode: pin mode (edge or state). The configuration have to be compatible.
+  * @retval None
+  */
+void LowPower_EnableWakeUpPin(uint32_t pin, uint32_t mode) {
+#ifndef STM32L4xx
+  UNUSED(mode);
+#endif
+  PinName p = digitalPinToPinName(pin);
+  uint32_t wupPin = pinmap_find_function(p, PinMap_WKUP);
+  switch (wupPin) {
+#ifdef PWR_WAKEUP_PIN1
+    case 1 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN1_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN1_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN1);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN1
+#ifdef PWR_WAKEUP_PIN2
+    case 2 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN2_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN2_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN2);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN2
+#ifdef PWR_WAKEUP_PIN3
+    case 3 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN3_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN3_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN3);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN3
+#ifdef PWR_WAKEUP_PIN4
+    case 4 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN4
+#ifdef PWR_WAKEUP_PIN5
+    case 5 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN5
+    default :
+      break;
+  }
+}
+
+/**
+  * @brief  Enable the sleep mode.
+  * @param  None
+  * @retval None
+  */
+void LowPower_sleep(){
+  /*Suspend Tick increment to prevent wakeup by Systick interrupt.
+    Otherwise the Systick interrupt will wake up the device within 1ms (HAL time base)*/
+  HAL_SuspendTick();
+
+  /* Enter Sleep Mode , wake up is done once User push-button is pressed */
+  HAL_PWR_EnterSLEEPMode(PWR_MAINREGULATOR_ON, PWR_SLEEPENTRY_WFI);
+
+  /* Resume Tick interrupt if disabled prior to SLEEP mode entry */
+  HAL_ResumeTick();
+
+  if (WakeUpUartCb != NULL) {
+    WakeUpUartCb();
```

Comment:
this is not clear why UART callback is always called, even if wake-up source is not UART ?

Reply:
At first sight, we aren't able to know which source wake up the MCU. Maybe we have miss something.